File: 3. Linked List/11_insert_node.java
class Node<T> {
    T val;
    Node<T> next;

    public Node(T val) {
        this.val = val;
        this.next = null;
    }
}

class Source {
    public static <T> Node<T> insertNode(Node<T> head, T value, int index) {

        Node<T> newNode = new Node<T>(value);

        // edge case -> insert before head
        if (index == 0) {
            newNode.next = head;
            // new head of linked list
            return newNode;
        }

        int count = 0;
        Node<T> current = head.next;
        Node<T> prev = null;

        while (current != null) {
            count++;
            if (index == count) {
                newNode.next = current;
                prev.next = newNode;
                break;
            }
            prev = current;
            current = current.next;
        }

        // edge case -> insert at the end
        if (index > count) {
            prev.next = newNode;
        }

        return head;
    }
}

File: 3. Linked List/11_insert_node/11_insert_node.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def insert_node(head, value, index):
    new_node = Node(value)

    if index == 0:
        new_node.next = head
        return new_node

    count = 0
    current = head.next
    prev = None

    while current is not None:
        count += 1
        if index == count:
            new_node.next = current
            prev.next = new_node
            break
        prev = current
        current = current.next

    if index > count:
        prev.next = new_node

    return head
